django/cantusdb_project/main_app/permissions.py

```diff
@@ -0,0 +1,140 @@
+from django.db.models import Q
+
+
+def user_can_edit_chants_in_source(user, source):
+    """
+    Checks if the user can edit Chants in a given Source.
+    Used in ChantDetail, ChantList, ChantCreate, ChantDelete, ChantEdit,
+    ChantEditSyllabification, and SourceDetail views.
+    """
+    if user.is_anonymous or (source is None):
+        return False
+
+    source_id = source.id
+    user_is_assigned_to_source = user.sources_user_can_edit.filter(id=source_id)
```

Comment:
`user.sources_user_can_edit.filter(id=source_id).exists()` would be clearer.

We could annotate variables throughout this PR. `user_is_assigned_to_source: queryset = user.sources_user_can_edit.filter(id=source_id)` is an accurate description of what this variable contains, but it also reveals that the variable's name doesn't reflect what's there. Better would be `user_is_assigned_to_source: bool = user.sources_user_can_edit.filter(id=source_id).exists()`